### PR TITLE
[feature](vector-search) add 4 distance functions to support vector search

### PR DESCRIPTION
**Pull request by cambyzju**
## Proposed changes

Issue Number: (https://github.com/apache/doris/issues/23128)

Add 4 distance functions to support brute-force search for vectors or embeddings
- DOUBLE l2_distance(vector1, vector2)
- DOUBLE inner_product(vector1, vector2)
- DOUBLE cosine_distance(vector1, vector2)
- DOUBLE l1_distance(vector1, vector2)

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by cambyzju** [comment]
run buildall

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 46.55 seconds
 stream load tsv:          538 seconds loaded 74807831229 Bytes, about 132 MB/s
 stream load json:         20 seconds loaded 2358488459 Bytes, about 112 MB/s
 stream load orc:          65 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          31 seconds loaded 861443392 Bytes, about 26 MB/s
 insert into select:          29.7 seconds inserted 10000000 Rows, about 336K ops/s
 storage size: 17162457150 Bytes

**Comment by cambyzju** [comment]
run buildall

**Comment by cambyzju** [comment]
run buildall

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 45.96 seconds
 stream load tsv:          535 seconds loaded 74807831229 Bytes, about 133 MB/s
 stream load json:         20 seconds loaded 2358488459 Bytes, about 112 MB/s
 stream load orc:          64 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          30 seconds loaded 861443392 Bytes, about 27 MB/s
 insert into select:          29.3 seconds inserted 10000000 Rows, about 341K ops/s
 storage size: 17162116937 Bytes

File: docs/zh-CN/docs/sql-manual/sql-functions/distance-functions/cosine_distance.md [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+under the License.
+-->
+
+## cosine_distance
+
+### description
+#### Syntax
+
+```sql
+DOUBLE cosine_distance(vector1, vector2)
```
[the remark this answers is not in the document]
Comment (cambyzju) [reply docs/zh-CN/docs/sql-manual/sql-functions/distance-functions/cosine_distance.md]: done

File: docs/zh-CN/docs/sql-manual/sql-functions/distance-functions/inner_product.md [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+```sql
+DOUBLE inner_product(vector1, vector2)
+```
+
+计算两个大小相同的向量的标量积
+
+### example
+
+```
+sql> SELECT cosine_distance([1, 2], [2, 3]);
```
[the remark this answers is not in the document]
Comment (cambyzju) [reply docs/zh-CN/docs/sql-manual/sql-functions/distance-functions/inner_product.md]: done

File: docs/en/docs/sql-manual/sql-functions/distance-functions/l2_distance.md [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+under the License.
+-->
+
+## l2_distance
+
+### description
+#### Syntax
+
+```sql
+DOUBLE l2_distance(vector1, vector2)
```
[the remark this answers is not in the document]
Comment (cambyzju) [reply docs/en/docs/sql-manual/sql-functions/distance-functions/l2_distance.md]: done

File: gensrc/script/doris_builtins_functions.py [hunk 4]
```diff
@@ -114,6 +114,38 @@
 
         [['element_at', '%element_extract%'], 'T', ['ARRAY<T>', 'BIGINT'], 'ALWAYS_NULLABLE', ['T']],
 
+        [['l1_distance'], 'DOUBLE', ['ARRAY_TINYINT', 'ARRAY_TINYINT'], 'ALWAYS_NULLABLE'],
```
[the remark this answers is not in the document]
Comment (cambyzju) [reply gensrc/script/doris_builtins_functions.py:117]: we use get_float64 to convert the type, before calculate it.

File: docs/zh-CN/docs/sql-manual/sql-functions/distance-functions/cosine_distance.md [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+
+```sql
+DOUBLE cosine_distance(ARRAY<T> array1, ARRAY<T> array2)
+```
+
+计算两个向量（向量值为坐标）之间的余弦距离
+如果输入array为NULL，或者array中任何元素为NULL，则返回NULL
+
+#### Notice
+* 输入数组的子类型支持：TINYINT、SMALLINT、INT、BIGINT、LARGEINT、FLOAT、DOUBLE
```
Comment (luozenglin) [inline docs/zh-CN/docs/sql-manual/sql-functions/distance-functions/cosine_distance.md:38]: Does nested type support decimal?

**Review by xy720 — APPROVED** [review]
LGTM

File: docs/zh-CN/docs/sql-manual/sql-functions/distance-functions/cosine_distance.md (hunk 5 above)
Comment (cambyzju) [reply docs/zh-CN/docs/sql-manual/sql-functions/distance-functions/cosine_distance.md:38]: Do not support now, actually vector<float> is enough.

**Comment by cambyzju** [comment]
run buildall

**Review by luozenglin — APPROVED** [review]
LGTM

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 46.69 seconds
 stream load tsv:          544 seconds loaded 74807831229 Bytes, about 131 MB/s
 stream load json:         20 seconds loaded 2358488459 Bytes, about 112 MB/s
 stream load orc:          65 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          31 seconds loaded 861443392 Bytes, about 26 MB/s
 insert into select:          29.7 seconds inserted 10000000 Rows, about 336K ops/s
 storage size: 17162154497 Bytes

**Comment by cambyzju** [comment]
run p0

**Pull request merged**